presto-kafka/src/main/java/io/prestosql/plugin/kafka/encoder/raw/RawRowEncoder.java

```diff
@@ -0,0 +1,334 @@
+            checkFieldType(columnHandle, fieldType);
+
+            valueLengths[i] = parseMapping(i, fieldType);
+            capacity += valueLengths[i];
+
+            checkArgument(isSupportedType(columnHandle.getType()), "Unsupported column type '%s' for column '%s'", columnHandle.getType().getDisplayName(), columnHandle.getName());
+        }
+
+        this.buffer = ByteBuffer.allocate(capacity);
+    }
+
+    private FieldType parseFieldType(EncoderColumnHandle columnHandle)
+    {
+        try {
+            return Optional.ofNullable(columnHandle.getDataFormat())
+                    .map(dataFormat -> FieldType.valueOf(dataFormat.toUpperCase(Locale.ENGLISH)))
+                    .orElse(FieldType.BYTE);
+        }
+        catch (IllegalArgumentException e) {
+            throw new IllegalArgumentException(format("Invalid dataFormat '%s' for column '%s'", columnHandle.getDataFormat(), columnHandle.getName()));
+        }
+    }
+
+    private void checkFieldType(EncoderColumnHandle columnHandle, FieldType fieldType)
+    {
+        String columnName = columnHandle.getName();
+        Type columnType = columnHandle.getType();
+        if (columnType == BIGINT) {
+            checkFieldTypeOneOf(fieldType, columnName, columnType, FieldType.BYTE, FieldType.SHORT, FieldType.INT, FieldType.LONG);
+        }
+        else if (columnType == INTEGER) {
+            checkFieldTypeOneOf(fieldType, columnName, columnType, FieldType.BYTE, FieldType.SHORT, FieldType.INT);
+        }
+        else if (columnType == SMALLINT) {
+            checkFieldTypeOneOf(fieldType, columnName, columnType, FieldType.BYTE, FieldType.SHORT);
+        }
+        else if (columnType == TINYINT) {
+            checkFieldTypeOneOf(fieldType, columnName, columnType, FieldType.BYTE);
+        }
+        else if (columnType == BOOLEAN) {
+            checkFieldTypeOneOf(fieldType, columnName, columnType, FieldType.BYTE, FieldType.SHORT, FieldType.INT, FieldType.LONG);
+        }
+        else if (columnType == DOUBLE) {
+            checkFieldTypeOneOf(fieldType, columnName, columnType, FieldType.DOUBLE, FieldType.FLOAT);
+        }
+        else if (isVarcharType(columnType)) {
+            checkFieldTypeOneOf(fieldType, columnName, columnType, FieldType.BYTE);
+        }
+    }
+
+    private void checkFieldTypeOneOf(FieldType declaredFieldType, String columnName, Type columnType, FieldType... allowedFieldTypes)
+    {
+        if (!Arrays.asList(allowedFieldTypes).contains(declaredFieldType)) {
+            throw new IllegalArgumentException(format(
+                    "Wrong dataFormat '%s' specified for column '%s'; %s type implies use of %s",
+                    declaredFieldType.name(),
+                    columnName,
+                    columnType.getDisplayName(),
+                    Joiner.on("/").join(allowedFieldTypes)));
+        }
+    }
+
+    private int parseMapping(int index, FieldType fieldType)
```

Comment:
I'll try this out, already sounds much easier than what I was doing.

Reply:
This still leaves the question of what to do with `VARCHAR` types that don't have an end specified. The way I see it we could,

1. Set the length equal to 1 (byte)
2. Guess the length based on the start mapping of the next column
3. Guess the length based on the length of the `VARCHAR` column type instance

Otherwise I don't know how many bytes to allocate to the `ByteBuffer`

Reply:
I think that if INSERT is to be supported for table we should require `end` to be specified for VARCHAR columns. Possibly allowing only exception where VARCHAR column maps to tail of message byte buffer; I am not sure if that is necessary though.
I think this is fine that INSERT not necessarily works for all table definitions for which SELECT works (overlapping fields is a good example where we cannot make INSERT work). 
Let's add constraints and focus on on basic scenarios first (with well defined semantics). Also let's make sure that for those, INSERT is compatible with SELECT (we can do roundtrip). If we see that it would be useful, we can relax contraints later.

Reply:
Based on reading the docs and skimming through the code, it seems to me that a `VARCHAR` mapping without an `end` specified is valid **only** for the last column in the mapping.  For any other position, I think we need the `end` index and we need to ensure that the length of the value we are inserting matches the size of the mapping.

I agree about the importance of round-trip tests, but we also need to make sure we conform to the spec we created originally, as up to this point, data has been created by non-Presto clients.